File: sdks/typescript/tests/transport-auth.test.ts
import { describe, expect, it } from "vitest";
import { AmqpTransportClient } from "../src/amqpTransport.js";
import { MqttTransportClient } from "../src/mqttTransport.js";
import { defaultAgentOptions, optionsFromConfigMap, optionsToConfigMap } from "../src/options.js";
import { httpAuthHeaders, parseAuthConfig } from "../src/transportAuth.js";

describe("Transport auth", () => {
  it("builds bearer authorization headers", () => {
    const auth = parseAuthConfig({
      type: "bearer",
      parameters: { token: "demo-token" }
    });
    expect(httpAuthHeaders(auth)).toEqual({ Authorization: "Bearer demo-token" });
  });

  it("embeds auth config in AMQP service hints", () => {
    const hint = AmqpTransportClient.buildServiceHint(
      "agent:sender@demo",
      "amqps://broker.local",
      "acp.exchange",
      {
        type: "username_password",
        parameters: { username: "agentA", password: "secret" }
      }
    );
    expect((hint.auth as { type: string }).type).toBe("username_password");
  });

  it("embeds auth config in MQTT service hints", () => {
    const hint = MqttTransportClient.buildServiceHint(
      "agent:sender@demo",
      "mqtts://broker.local:8883",
      undefined,
      1,
      "acp/agent",
      {
        type: "username_password",
        parameters: { username: "agentA", password: "secret" }
      }
    );
    expect((hint.auth as { type: string }).type).toBe("username_password");
  });

  it("parses direct and relay transport auth from options config map", () => {
    const options = optionsFromConfigMap({
      direct_transport_auth: {
        type: "bearer",
        parameters: { token: "direct-token" }
      },
      relay_transport_auth: {
        type: "bearer",
        parameters: { token: "relay-token" }
      }
    });
    expect(options.direct_transport_auth?.type).toBe("bearer");
    expect(options.relay_transport_auth?.parameters.token).toBe("relay-token");
  });

  it("serializes broker auth in options config map", () => {
    const exported = optionsToConfigMap({
      ...defaultAgentOptions(),
      amqp_auth: {
        type: "username_password",
        parameters: { username: "agentA", password: "secret" }
      },
      mqtt_auth: {
        type: "username_password",
        parameters: { username: "agentA", password: "secret" }
      }
    });
    expect((exported.amqp_auth as { type: string }).type).toBe("username_password");
    expect((exported.mqtt_auth as { type: string }).type).toBe("username_password");
  });
});

File: sdks/java/src/test/java/org/acp/client/TransportAuthTest.java
package org.acp.client;

import org.junit.jupiter.api.Test;

import java.util.Map;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotNull;

class TransportAuthTest {
    @Test
    void buildsBearerAuthHeaders() {
        AuthConfig auth = TransportAuth.parseAuthConfig(
            Map.of("type", "bearer", "parameters", Map.of("token", "demo-token"))
        );
        Map<String, String> headers = TransportAuth.httpAuthHeaders(auth);
        assertEquals("Bearer demo-token", headers.get("Authorization"));
    }

    @Test
    void embedsAuthInAmqpAndMqttServiceHints() {
        Map<String, Object> amqpHint = AmqpTransportClient.buildServiceHint(
            "agent:sender@demo",
            "amqps://broker.local",
            "acp.exchange",
            Map.of(
                "type",
                "username_password",
                "parameters",
                Map.of("username", "agentA", "password", "secret")
            )
        );
        assertEquals("username_password", asMap(amqpHint.get("auth")).get("type"));

        Map<String, Object> mqttHint = MqttTransportClient.buildServiceHint(
            "agent:sender@demo",
            "mqtts://broker.local:8883",
            null,
            1,
            "acp/agent",
            Map.of(
                "type",
                "username_password",
                "parameters",
                Map.of("username", "agentA", "password", "secret")
            )
        );
        assertEquals("username_password", asMap(mqttHint.get("auth")).get("type"));
    }

    @Test
    void parsesAndSerializesTransportAuthFromAgentOptions() {
        AcpAgentOptions options = new AcpAgentOptions()
            .setDirectTransportAuth(Map.of("type", "bearer", "parameters", Map.of("token", "direct-token")))
            .setRelayTransportAuth(Map.of("type", "bearer", "parameters", Map.of("token", "relay-token")))
            .setAmqpAuth(Map.of(
                "type",
                "username_password",
                "parameters",
                Map.of("username", "agentA", "password", "secret")
            ))
            .setMqttAuth(Map.of(
                "type",
                "username_password",
                "parameters",
                Map.of("username", "agentA", "password", "secret")
            ));
        Map<String, Object> exported = options.toConfigMap();
        assertNotNull(exported.get("direct_transport_auth"));
        assertNotNull(exported.get("relay_transport_auth"));
        assertNotNull(exported.get("amqp_auth"));
        assertNotNull(exported.get("mqtt_auth"));

        AcpAgentOptions parsed = AcpAgentOptions.fromConfigMap(exported);
        assertEquals("bearer", asMap(parsed.getDirectTransportAuth()).get("type"));
        assertEquals("bearer", asMap(parsed.getRelayTransportAuth()).get("type"));
        assertEquals("username_password", asMap(parsed.getAmqpAuth()).get("type"));
        assertEquals("username_password", asMap(parsed.getMqttAuth()).get("type"));
    }

    @SuppressWarnings("unchecked")
    private static Map<String, Object> asMap(Object value) {
        return (Map<String, Object>) value;
    }
}
